core/src/main/java/io/grpc/internal/ManagedChannelImpl.java

```diff
@@ -527,7 +536,43 @@ public boolean isTerminated() {
     return terminated;
   }
 
-  /*
+  @Override
+  public ConnectivityState getState(boolean requestConnection) {
+    // TODO: handle requestConnection. At this point, not really sure how.
+    return state;
+  }
+
+  @Override
+  public void notifyWhenStateChanged(ConnectivityState source, Runnable callback) {
```

Comment:
I think you're mis-understanding the semantics of `source`. There is no need to remember `source` after an initial check.

`source` is the currently-believed state, to handle racing. So if `source != state`, then `callback` is run immediately. After that point though, it will be the same as the current state, so you only need to call all `callback`s when the current state changes.